### PR TITLE
Remove box syntax from compiler and tools

### PR DESCRIPTION
**Pull request by est31**
Removes box syntax from the compiler and tools. In #49733, the future of box syntax is uncertain and the use in the compiler was listed as one of the reasons to keep it. Removal of box syntax [might affect the code generated](https://github.com/rust-lang/rust/pull/49646#issuecomment-379219615) and slow down the compiler so I'd recommend doing a perf run on this.

### CODE REVIEW

**Comment by jyn514** [comment]
https://github.com/rust-lang/rust/issues/49733 says

> how easy is it to remove box syntax as a language feature? Can we just describe it as "will never be stabilised in its current form as it's an internal implementation detail, may be stabilised under a new design of placement")? 
> is there any reason for users to use box syntax (apart from it looking nice) given Box::new is inline always, or does that not work?
> if we decide to keep this, there are a number of unresolved questions in box and in expressions (tracking issue for RFC 809) #22181 about the syntax behaviour

That sounds to me like it's not at all certain that it's going to be removed.

**Comment by est31** [comment]
@jyn514 good point but note that the issue also contains:

> I didn't want to remove it (yet) due to widespread usage in the compiler (Box::new is implemented in terms of it) and in the wider ecosystem (unlike <- etc which only had a handful of crates using it).

This PR alleviates the "widespread usage in the compiler" point, if wanted.

Also note that currently the compiler has inconsistent usage of `Box::new` vs `box`. Currently if I grep for `Box::new` in the compiler/ dir, I get 278 uses. With this PR, there would be 413 uses. So it's 135 uses of `box` syntax, or 33% of all box constructions. If this PR turns out to have a significant negative performance impact, it won't just influence the decision on whether to keep the `box` language feature, one could think about introducing the box syntax to more places in the compiler, in the hope of speeding up things.

**Comment by jyn514** [comment]
@rust-timer queue

**Comment by est31** [comment]
(also I have edited the PR description to remove the claim that the tracking issue plans certain removal)

**Comment by hellow554** [comment]
One question though: Without the `box` feature, how will [`Box::new`](https://github.com/rust-lang/rust/blob/d4ad1cfc63ba5824196bfb2370451ddb5af2e020/library/alloc/src/boxed.rs#L191-L193) be implemented in the future? Will it become a language item?

**Comment by leonardo-m** [comment]
>So it's 135 uses of box syntax, or 33% of all box constructions. If this PR turns out to have a significant negative performance impact,

Another benchmark worth doing is to turn all of them into box syntax.

**Comment by bjorn3** [comment]
> One question though: Without the box feature, how will Box::new be implemented in the future? Will it become a language item?

It could call the allocator without lang item indirection like it currently does (`exchange_malloc`), use `ptr::write` to store the value and then construct the `Box` directly from the pointer. All `(try_)new_*` methods already need to do this instead of `box expr`.

**Comment by hellow554** [comment]
If I see it correctly, you haven't removed every occurence of the `box`, right?

I really would like to see that. I think it would be easiest to remove the features `box_syntax` and `box_patterns` and see where the compiler complains to get rid of everything.

**Comment by jyn514** [comment]
Please do not remove box_patterns. It makes code vastly more easy to read and it also has a potential replacement with the Deref pattern project.

**Comment by hellow554** [comment]
Can we have another timer run?

**Comment by nbdd0121** [comment]
`Box::new` needs a move while `box` can construct in-place. The MIR for `Box::new` would be something like (pseudo-code)
```
_1 = construct the value
_2 = construct the box
(*_2) = move _1
```
while the box syntax gives
```
_1 = construct the box;
(*_1) = construct the value
```

And the reality this is more complicated if the `Box::new` is to be implemented by method calls to `exchange_malloc` and `write` (which is not itself a intrinsic) etc. `try_new` method is different that it cannot be assumed to be successful so for semantics you'll need to construct the value first and move it later anyway.

Can we be 100% certain with this change that values will still be constructed in place for things like `RcBox`?

**Comment by est31** [comment]
@hellow554 if I check the library dir, there seem to be a lot of uses of box syntax still in the tests and stuff. Maybe make a new PR to remove it from there as well?

**Comment by nbdd0121** [comment]
I digged into some of the previously poorly optimised cases (https://godbolt.org/z/o6a9b3K5M), and it seems that they are properly optimised in Rust 1.52 (while not in Rust 1.51).

Rust 1.52 updates LLVM to 12, so I guess with LLVM 12, `box` would no longer have performance advantage given most calls to `Box::new` are properly optimised. We still support LLVM 10 and 11 though. Do we have a way to do a perf run with a different LLVM version?

**Comment by est31** [comment]
Thanks for digging into this @nbdd0121 ! About the LLVM versions, one could probably turn back the time and use a different LLVM version in this PR and then re-run a perf run.

I agree that older LLVM versions should be fully supported, even if this means a delay for box syntax removal by a year or two. There is no rush and the distros that use old LLVM versions will be thankful. That being said, one could also do this via a dedicated MIR optimization instead of relying on LLVM. This wouldn't just benefit `Box::new` but any other function that looks like it.

**Comment by nbdd0121** [comment]
A MIR pass that turns `Box::new` into `box` does sound like a good idea (and would hopefully saves some compilation time as less work is required by LLVM).

**Comment by est31** [comment]
Thats's a lot of red wow...

Screenshot_20210807_041429

I guess it doesn't like the other `Box::new` implementation that uses `Box::new_in`? That's the biggest difference I can see to the prior perf run...

**Comment by hellow554** [comment]
> @hellow554 there seem to be a lot of uses of box syntax still in the tests and stuff

Yes, I haven't touched ans Tests at all. First because there are a lot of regression tests that I would like to be untouched and second I thought they wouldn't make any difference in this case, do they?

That's really a big regression 😨
Can we try to find out why? Is it really `Box::new` that causes it?

**Comment by est31** [comment]
> Can we try to find out why? Is it really Box::new that causes it?

We ran a perf run before your commit and one after it. The perf run before your commit shows no performance changes, while the one after it shows significant performance changes. Outside of `Box::new`, I can't see any other changes in that commit that could have caused such large regressions.

There are two ways to test this hypothesis. First, making a fresh PR with only the `Box::new` change. However, that will cause a much smaller regression because currently, large parts of the compiler are still using box syntax, which this PR has changed, so you'd be getting different numbers. The second way to test it would be to take this PR, revert the `Box::new` change, and keep everything else unchanged and do another perf run. I think this is the best way forward.

**Comment by nbdd0121** [comment]
I just created a commit that will converts `Box::new` into `box` during THIR -> MIR building (https://github.com/nbdd0121/rust/commit/4e678bb81341cf3e16c7df5a6025bbe2124bfd6e). Maybe we can try a perf run with that commit added?

**Comment by est31** [comment]
@nbdd0121 thanks for that commit, that was really quick! I'll try to check if I can apply it to this PR. It's not trivial because it's changing code that uses `box` syntax itself. Rebasing my changes onto that commit seems easier I think.

**Comment by nbdd0121** [comment]
It's actually an indentation change that causes conflict. I guess you can just `git checkout --theirs <conflict>` and re-remove `box`.

**Comment by nbdd0121** [comment]
It seems that my commit when interacting with two-phase borrow actually causes ICE. This might be the correct way 😢

**Comment by nbdd0121** [comment]
Borrow checker ICE is fixed in https://github.com/nbdd0121/rust/commit/ca0b1c74c2036e9f3747afa7541d863f2ff692ac. 
```
// Without `Box::new` to `box` transform
_1 = construct the value;
_2 = Box::new(move _1);

// With previous commit or box syntax
// but apparently `Box::new` syntax implies two-phase borrow and the borrow checker requires construct into locals
_1 = Box;
(*_1) = construct the value;

// With the fix
_1 = Box;
_2 = construct the value;
(*_1) = *2;
```

**Comment by est31** [comment]
@jyn514 could you issue another perf run? Thanks.

**Comment by bjorn3** [comment]
@rust-timer check 4b9300307cb984bee4668db59cd24167bd5b4ac8

**Comment by bjorn3** [comment]
Wrong command.

@rust-timer build 4b9300307cb984bee4668db59cd24167bd5b4ac8

**Comment by jyn514** [comment]
@rust-timer build https://github.com/rust-lang/rust/commit/4b9300307cb984bee4668db59cd24167bd5b4ac8

**Comment by est31** [comment]
Huh, still red, although way less red than before.

Screenshot_20210809_004842

The test that is the most red is the [`deep-vector`](https://github.com/rust-lang/rustc-perf/blob/cfea1251520dcb69147df8965048a935fdf1eefe/collector/benchmarks/deep-vector/src/main.rs) one. It constructs a huge vector made up of lots of zeros. Among other changes, this PR alters the `vec` macro to use `Box::new` instead of box syntax. Maybe the MIR pass doesn't work and it still constructs the giant array on the stack? The big regressions of `deep-vector-opt` are in the LLVM passes. In the other regressing examples, I couldn't find a single responsible pass.

**Comment by nbdd0121** [comment]
Refer to my previous comment, due to borrow checker dislike direct construction into the box, I still need to construct on a local temporary (_2) first before moving it into `*_1`. With no local copy propagation pass currently available in MIR this means that it's a large copy that LLVM has to optimise out.

**Comment by est31** [comment]
@nbdd0121 I see that would explain it. I understand why two phase borrows would want parameters in locals first. I'm not familiar with the borrow checker code, but from my understanding, two phase borrows are only needed for functions taking `&mut self`, no? So maybe this requirement that ICEs if the param is not available as a local could be relaxed to only require it for functions taking `&mut self`?

**Comment by leonardo-m** [comment]
If I replace some box [somearray] with Box::new([somearray]) in my code I get stack overflows. Will this be fixed before the box syntax get removed?

**Comment by est31** [comment]
The PR is [blocked](https://github.com/rust-lang/rust/pull/87781#issuecomment-894872854) right now on improving the pass that replaces `Box::new` with something even more equivalent to `box` syntax.

That being said, the state before the library changes had no observable regressions, see the [first perf run](https://github.com/rust-lang/rust/pull/87781#issuecomment-893214467). If someone wants me to revert the PR to that old state and wants to approve it for merge, please tell me.

**Comment by oli-obk** [comment]
> If someone wants me to revert the PR to that old state and wants to approve it for merge, please tell me.

i'd r+ that. people may reintroduce it, but doing it again later will be much easier with the big chunk already done

**Comment by nbdd0121** [comment]
A friendly reminder that this'll still like to regress LLVM 10 and 11.

Also it might worth adding a few codegen tests for `Box::new` to ensure that those optimisations won't regress for previously badly optimized cases (like https://godbolt.org/z/o6a9b3K5M that I mentioned earlier).

**Comment by oli-obk** [comment]
> A friendly reminder that this'll still like to regress LLVM 10 and 11.

So, compilers built with these llvm versions will be slower after this PR? It won't affect user code, as it's just compiler internal crate changes.

**Comment by est31** [comment]
@nbdd0121 oh right I forgot about that. Yeah then the MIR optimization pass would be needed, or we wait. It's hard to measure how much removal of box syntax is slowing down compilers compiled with old LLVM versions, as it is hard to measure how much the MIR pass is helping with that.

**Comment by nbdd0121** [comment]
Sorry, I thought that there are still library changes (probably the title should be changed).

So @oli-obk is correct, this only slows compilers built with the old LLVM version and wouldn't affect users. I suppose that wouldn't be a problem then.

**Comment by oli-obk** [comment]
@bors r+ rollup=never

Ok, let's merge it then. The older llvm versions will get their compiler perf back once we have the mir opt

**Comment by est31** [comment]
@oli-obk your prediction that new boxes will be added turned out to be right faster than anticipated :laughing: . This PR's state is clean, basing on aa8f27bf4d980023a8b245ceb25a490a18041eb2. But already the merge with  30a0a9b694cde95cbab863f7ef4d554f0f46b606 caused issues, namely with ad2b4f44412391ff14d96d71d7b6a212f1d94755 that added a `box` occurence. I've fixed it and re-pushed.

File: compiler/rustc_codegen_cranelift/example/alloc_example.rs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#![feature(start, box_syntax, core_intrinsics, alloc_prelude, alloc_error_handler)]
+#![feature(start, core_intrinsics, alloc_prelude, alloc_error_handler)]
```
Comment (bjorn3) [inline compiler/rustc_codegen_cranelift/example/alloc_example.rs:1]: These actually do still use the box syntax. I guess I will need to update them at the next sync.

Comment (est31) [reply compiler/rustc_codegen_cranelift/example/alloc_example.rs:1]: Oh sorry about that. CI isn't ran for these so it didn't detect it but I could have double checked.

**Comment by est31** [comment]
@nbdd0121 if you want to submit your commits as a PR, I've rebased them and they are available on my [remove_box](https://github.com/est31/rust/tree/remove_box) branch. cheers!

**Comment by est31** [comment]
Followup PR for the testsuite: #88316

For the question of removing box syntax support from the compiler, there is not much of a difference if box syntax is used in the testsuite or the compiler itself. So I have made a followup PR to remove box syntax from there too, at least outside of the issues directory.

**Comment by jyn514** [comment]
> So I have made a followup PR to remove box syntax from there too, at least outside of the issues directory.

This seems like a really bad idea. How will you possibly ensure that it's testing the same thing that it did before?

**Comment by Mark-Simulacrum** [comment]
Was going to remove the perf-regression, but it looks like the final merge of this PR did actually results in a slight [regression](https://perf.rust-lang.org/compare.html?calcNewSIg=true&start=896f058f13d6c8021f7637817953a44d3a78be32&end=ba8cda2fa2c99ed6646f4dfe73bf4edad7e42a2d&stat=instructions%3Au), primarily for -doc builds. My guess is that some of those larger types in rustdoc are being copied a little more than previously.

**Comment by est31** [comment]
Hmm yeah maybe the rustdoc part should be reverted? I can do a PR for it.

**Comment by est31** [comment]
Or we can wait until #89030 merges.

**Comment by nbdd0121** [comment]
I think regression isn't bad enough to warrant a reversion, especially since it's desirable to phase out box expression in the long term.

**Comment by est31** [comment]
Hmm a glance at #89030 makes it seem that `ShallowInitBox` only affects box syntax right? So it shouldn't improve `Box::new` calls like the commit "Transform Box::new to box" commit?

**Comment by nbdd0121** [comment]
No, it's not related to the transform. That PR is to allow `box` expression to unwind so that `oom=panic` support could be unblocked. (And it would in fact regress performance slightly for all box expressions)

**Comment by jyn514** [comment]
> Hmm yeah maybe the rustdoc part should be reverted? I can do a PR for it.

Please do, yes - collect_blanket_impls especially is absurdly performance sensitive and I don't think this is useful enough to regress the performance there.

**Comment by Mark-Simulacrum** [comment]
I don't think a straight revert is necessarily the right approach -- the regressions here are tiny, realistically, and [wall times](https://perf.rust-lang.org/compare.html?calcNewSig=true&start=896f058f13d6c8021f7637817953a44d3a78be32&end=ba8cda2fa2c99ed6646f4dfe73bf4edad7e42a2d&stat=wall-time) do not show a corresponding significant decrease.

Cachegrind [diff](https://gist.github.com/Mark-Simulacrum/cd6ea8880ec0dde9b599d0ad9684a565) for serde-doc point to a particular set of functions which likely drive the regressions -- targeted reverts likely make more sense here than a blanket revert. That said, if we revert everything in rustdoc for now that doesn't seem the end of the world either. Ultimately, this just points to Box::new being slightly less able to optimize than box, which was already well known.

**Comment by Mark-Simulacrum** [comment]
I also see no particular reason to believe that collect_blanket_impls has been affected by this patch -- can you say more about that perhaps?

**Comment by est31** [comment]
I've filed a PR in #89134 reverting every change of librustdoc. If anything can stay, please point it out!

**Comment by safinaskar** [comment]
https://github.com/rust-lang/rust/pull/87781#issuecomment-894532456 :
> try_new method is different that it cannot be assumed to be successful so for semantics you'll need to construct the value first and move it later anyway

?????
So, `try_new` always performs move? This is very bad. Compare this with C++ code:
```c++
try {
   a = new some_constructor(some_args);
}catch(...){
  //...
}
```
Here we can check for allocation errors while directly constructing value in place!

**Comment by bjorn3** [comment]
On nightly you can use `Box::new_uninit()` followed by `.assume_init()` after writing the value. This is unsafe though. On stable you can use `Box::new()` of a `Box<MaybeUninit<T>>` type and `transmute` to `Box<T>` after writing the value.

```rust
#![feature(new_uninit)]

let mut five = Box::<u32>::new_uninit();

let five: Box<u32> = unsafe {
    // Deferred initialization:
    five.as_mut_ptr().write(5);

    five.assume_init()
};
```

You can do something similar for the `try_*` variant.

**Pull request merged**